bccsp/pkcs11/pkcs11.go

```diff
@@ -51,46 +52,37 @@ func loadLib(lib, pin, label string) (*pkcs11.Ctx, uint, *pkcs11.SessionHandle,
 		}
 	}
 	if !found {
-		return nil, slot, nil, fmt.Errorf("Could not find token with label %s", label)
+		return nil, slot, nil, fmt.Errorf("could not find token with label %s", label)
 	}
 
-	session := createSession(ctx, slot, pin)
-
-	logger.Debugf("Created new pkcs11 session %+v on slot %d\n", session, slot)
-
-	if pin == "" {
-		return nil, slot, nil, fmt.Errorf("No PIN set")
-	}
-	err = ctx.Login(session, pkcs11.CKU_USER, pin)
+	session, err := createSession(ctx, slot, pin)
 	if err != nil {
-		if err != pkcs11.Error(pkcs11.CKR_USER_ALREADY_LOGGED_IN) {
-			return nil, slot, nil, fmt.Errorf("Login failed [%s]", err)
-		}
+		return nil, slot, nil, err
 	}
 
 	return ctx, slot, &session, nil
 }
 
-func (csp *impl) getSession() (session pkcs11.SessionHandle) {
+func (csp *impl) getSession() (session pkcs11.SessionHandle, err error) {
 	select {
 	case session = <-csp.sessions:
-		_, err := csp.ctx.GetSessionInfo(session)
+		_, err = csp.ctx.GetSessionInfo(session)
 		if err != nil {
 			logger.Warningf("Get session info failed [%s], closing existing session and getting a new session\n", err)
 			csp.ctx.CloseSession(session)
-			session = createSession(csp.ctx, csp.slot, csp.pin)
+			session, err = createSession(csp.ctx, csp.slot, csp.pin)
```

Comment:
Keep in mind this error will overwrite the one on line 69 for the purposes of the return on 82. Is there ever a case where we would not want to attempt creating a new Session if GetSessionInfo fails and we would want to return early? I'm guessing that's largely dependent on error handling specific errors from `GetSessionInfo`.

Reply:
No the err on line 69 is essentially acting as an `ok` flag. Its used internally only. The error propagated back to the caller should be the failure to create a new session, not the fact the the session was invalid at some point in the flow.